### ISSUE TITLE
Does not work with Firefox 

### ISSUE THREAD

**Issue by tutoduino** [polhenarejos/pico-fido#65]
Hello,
PicoFIDO fails to run on Raspberry Pi Pico with Firefox 131.0 under Linux Mint. 
Testing on https://www.token2.com/tools/fido2-demo
It works fine with Chromium, but I need this solution to work with Firefox...
Thanks for your support.

**Comment by polhenarejos** [comment]
Can you try flashing with a nightly development build?

**Comment by tutoduino** [comment]
I did, the result of registration is "operation cancelled" after PIN code is entered and button is pushed.
See attachment.
Thanks for your support.
![error](https://github.com/user-attachments/assets/9d47b44c-f8a1-4e68-a319-9a6a30634548)

**Comment by polhenarejos** [comment]
And it works well with Chrome, right?

**Comment by polhenarejos** [comment]
With Firefox in Windows 10 works. Not sure if it is a linux problem. However, compared to Chrome, Firefox is extremely slow. It takes almost 1 minute to perform the process, whereas Chrome takes 2-3 seconds.

**Comment by tutoduino** [comment]
Interesting thanks, then I tried on Firefox under Windows (I do not have Windows at home, so need to test another system :)) and indeed it works. 
I am exploring the permission under Linux, because when I run the FIDO2 Key Data Explorer in Firefox/Linux I get an user permission error... 
![error](https://github.com/user-attachments/assets/6284ea36-3766-4a47-b3aa-b224d3fee5ad)

**Comment by polhenarejos** [comment]
I did not succeed in macOS neither. Firefox is sending continuously a bogus AAID for an unknown reason, which hangs the pico fido until it is closed. Perhaps is related. 

If you are skilled, you can build the firmware with “DEBUG_APDU=1” in cmake and debuging UART0. I use a TTL to USB dongle that costs a couple of dollars using a terminal opened to the exposed COM port. I can review the log then.

**Comment by tutoduino** [comment]
On my setup (Firefox/Linux), my 2 other "real" FIDO2 key work fine.

**Comment by tutoduino** [comment]
Compilation fine, on which pins did you configure UART0 and at which speed ? 
Thanks

**Comment by polhenarejos** [comment]
Nothing else?

**Comment by tutoduino** [comment]
No

**Comment by tutoduino** [comment]
I register the key on Chrome successfuly (PIN is setup). And then I try again to register on Firefox. 
The first steps are fine, I enter the PIN, touch the key, then the message operation cancelled appears. 
Here are the traces in this case. 
[putty.log](https://github.com/user-attachments/files/17333391/putty.log)

**Comment by polhenarejos** [comment]
Does it work with https://webauthn.io ?
Did you test it with no cookies and blank pico fido? Withour resident keys. 

Perhaps this helps to debug

https://addons.mozilla.org/en-US/firefox/addon/webdevauthn/

**Comment by tutoduino** [comment]
I tested with no cookies and blank pico without resident keys. 
Interesting warning on webauthn.io, looks like permission issue !
I do not understand how to use the webdevauthn addon, where are displayed the debug information ?
![webauthn-io](https://github.com/user-attachments/assets/f0a46e99-8b5b-4a26-81b5-df01122fb5db)

**Comment by tutoduino** [comment]
I checked the forums and found https://webauthn.bin.coffee/  testing site that raises similar issue. 

![bin](https://github.com/user-attachments/assets/9982662b-d86d-4f9a-b91c-e717277493cb)

**Comment by polhenarejos** [comment]
Related with #46

**Comment by polhenarejos** [comment]
Does it work?

"Adding on to the previous answer by Cody, and the respective comments, you must also enable `security.webauth.u2f` in Firefox's `about:config` for U2F, FIDO2, and subsequently Yubikeys to work."

**Comment by tutoduino** [comment]
By default, firefox config set webauth flags correctly (flag names are different now, see attached picture). All flags set to true doe not work better.
![firefox-config](https://github.com/user-attachments/assets/2b708a3f-c51d-4022-b19c-270189a34d6f)

**Comment by tutoduino** [comment]
I see on some [forums](https://support.mozilla.org/mk/questions/1412073) that downloading Firefox from the official repository solved their issue.
I tried but it does not work for my problem here...

**Comment by polhenarejos** [comment]
And enabling direct attestation?

**Comment by tutoduino** [comment]
tried to enable all, does not work better

**Comment by tutoduino** [comment]
Firefox support could help ?

**Comment by polhenarejos** [comment]
What is strange to me is that your true Yubikey works smoothly with Firefox. There's something in it that is recognized by Firefox somehow.

**Comment by tutoduino** [comment]
Yes indeed. An I suspected udev but pico works fine with chromium...

**Comment by polhenarejos** [comment]
Any progress?

**Comment by tutoduino** [comment]
no idea of ​​research path...

**Comment by LinOx2** [comment]
I have the same issue with Firefox.
I use Falkon (chromium) for register, after I can use Firefox (131.0.3) for login, it work like that for me.

Falkon (webkit) on linux. Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Falkon/22.04.3 Chrome.

so Firefox cancel only register. BUGs in Firefox ???.

**Comment by tutoduino** [comment]
Firefox works fine with other FIDO2 keys so unclear to me...

**Comment by LinOx2** [comment]
I have Linux Mint and the same issue, and I use Raspberry Pi Pico (rp2040) in the way I have explain.
firefox reject (cancel) the pico key at register, but work well at login, if i have register with Falkon the pico key.
Maybe reading the Firefox code on fido register could explain what the problem is.

**Comment by patvdleer** [comment]
I can confirm I have the same issues, I'm using the Waveshare RP2040 One running on `Ubuntu 22.04.4 LTS` with `6.8.0-45-generic`.  I have the `Yubico NEO OTP+U2F+CCID` and the `SoloKeys Solo 4.0.0` which do work both on Firefox and Chrome.

Since I bought multiple waveshares I have one with and one without a pin set, both don't work.

Any way to further debug this?

Firefox settings:
![image](https://github.com/user-attachments/assets/d5e0910c-2882-46e0-b1c5-ef9dc65c3183)

**Comment by polhenarejos** [comment]
So which is the combination that does not work?

**Comment by patvdleer** [comment]
On Linux/Ubuntu on firefox the picokey doesnt work, others do. The picokey does work on Chrome.

On Mac it does work on Firefox and Chrome.

**Comment by polhenarejos** [comment]
I suspect it is related with the CCID interface. Firefox sends APDUs trying to find a specific file reference, but I do not know which one is the good one present in Yubikeys or Solo. 

If you want to debug, you can try to generate a log file of pcscd process. More info at https://ccid.apdu.fr/ (see Log section). Do it with your Yubikey. Plug the Yubikey and start Firefox. 
In the log file we’ll see the communication exchanged with the Yubikey and Firefox and I hope we’ll find which are the references needed by Firefox.

**Comment by air-eat** [comment]
i don't know if this is related, but disabling `security.webauthn.ctap2` for me seems to make it pass [various](https://demo.yubico.com/webauthn-technical) [tests](https://webauthn.bin.coffee/), and registering/signing in, although not through the "Or, sign in with passkey option", works with discord (even though none of the examples ask for the pin or recognize the device)

however, it does fail immediately in other areas like github passkey registration

**Comment by air-eat** [comment]
btw, i get the same behaviour with firefox on both macos and linux

**Comment by leecheeyong** [comment]
I'm not sure if this is related.

So I had everything working well before updating to macOS 18, on Firefox. Previously when the mac prompt the take action on the security key box, the Pico would flash on and off with a delay of 1 second but now when the box prompt, the pico will either keep the led on forever or kept off. It seems like the mac literally doesn't see it security key or something. I thought something was corrupted on my Pico, so I even tried re-flashing the firmware with the latest ones (version 6) but this time when it is plugged in, the pico wouldn't even blink and be detected for some reason. And then I tried going down version by version and ended up in version 3 where it will blink when plugged in but face the same issue as what happened initially.

On Windows however, it would be detected and blink as it should, but sometimes I'll have to reconnect the device multiple times for it to actually work as it would rapidly blink but does nothing sometimes.

**Comment by polhenarejos** [comment]
Can you try the nightly development build? There were some issues fixed recently.

**Comment by leecheeyong** [comment]
> Can you try the nightly development build? There were some issues fixed recently.

Yep, I've tried all of them, and it seems like its an issue on the mac itself but I don't know what it is

**Comment by air-eat** [comment]
yubikey fido test and github login still error out in the same way but now at least ask for a pin and presence auth, so it seems like a step in the right direction

**Comment by polhenarejos** [comment]
Can you check no NSS modules are interfering? I had a problem in the past with one PKCS11 module that stole the connection.

You can check it at Settings > Privacy & Security > Certificates > Security Devices 
I have only 3 enabled devices:
- NSS Internal
- Internal root
- OSSClient

**Comment by air-eat** [comment]
if i had any, it wouldn't explain the identical behavior across two different devices (pc with linux, laptop with macOS)

**Comment by polhenarejos** [comment]
Usually OS handle smartcards differently. macOS relies on libccid, Windows relies on scard driver. And despite pico-fido is mainly used by hid interface, it has a ccid interface that may lock the card if a daemon requires it. I had this problem with firefox in macOS. Whilst it worked in Windows it did not work with macOS, due to the PKCS11 module in firefox.

**Comment by randalthor17** [comment]
I can confirm the same issues. Firefox on linux doesnt work, both in flatpak and rpm. Chromium on linux works in both cases. Will test on macos and update.

**Comment by patvdleer** [comment]
I fixed it for me, I run the non snap version of Firefox and came across https://askubuntu.com/a/1520568/21245

I've added the following to `/etc/apparmor.d/usr.bin.firefox`
```
  # FIDO
  /sys/class/hidraw/ r,
  /dev/hidraw* rw,

  # Yubikey NEO also needs this:
  /sys/devices/**/hidraw/hidraw*/uevent r,
```

and reloaded apparmor
```bash
sudo service apparmor reload
```

**Comment by polhenarejos** [comment]
No activity. Reopen if needed.
